### ISSUE TITLE
Lingle-Clark bed deformation model: stopping and re-starting affects results

### ISSUE THREAD

**Issue by juseg** [pism/pism#370]
Potentially a major concern for anyone computing bed deformation in restarted runs.

Variable `topg_initial` is reset to `topg` at every restart (in stable v0.7.3).

    pisms -bed_def lc -o out1.nc
    pisms -bed_def lc -i out1.nc -o out2.nc
    ncview out?.nc

Do bed deformation models use a new equilibrium topography at every restart?
I had excpected PISM to go on with the same equilibrium topography as in the previous run.
Very worried about this one: most of my published and unpublished results are concerned.

Here is another example showing dbdt jumps at every restart.

    pisms -Lz 5000 -y 0 -o out0.nc
    for i in {0..7}
    do
        pisms -bed_def lc -i out$((i)).nc -o out$((i+1)).nc \
            -extra_file ex$((i)).nc -extra_times 100 \
            -extra_vars dbdt,topg,topg_initial
    done
    ncview ex?.nc

**Comment by talbrecht** [comment]
uff, the same problem for the dev branch! so the restart should look for topg_initial as for dbdt...

**Comment by talbrecht** [comment]
I looked a little bit into the code: topg_initial is used for the "special" topg correction for regridding!? That's why it is set to the old topg for every restart... I have assumed that topg, thk and dbdt at restart should provide a good approximation of the rebound state to restart from and eventually recover to the original (equilibrium) state topg0 for the original load thk0... However, in your example, topg varies by about 10% in out8.nc compared to the same simulation without intermediate restarts, and I do not find an explanation for it!?

**Comment by talbrecht** [comment]
I see small bumps in dbdt on the short time scale, but on the long time scale (90kyr) the bed actually recovers (practically) completely, no matter whether with or without (9) intermediate restarts... 

```
pisms -Lz 5000 -y 0 -o out_boot0.nc 

# run into equilibrium
pisms -Lz 5000 -y 90000 -o out_equi9a.nc \
-bed_def lc -i out_boot0.nc \
-extra_file ex_equi9a.nc -extra_times 1000 \
-extra_vars dbdt,topg,topg_initial,thk

# run into equilibrium with 9 restarts
cp out_boot0.nc out_equi0b.nc
for i in {0..8}
do
    pisms -bed_def lc -i out_equi$((i))b.nc -o out_equi$((i+1))b.nc \
    -extra_file ex_equi$((i+1))b.nc -extra_times 1000 -y 10000 \
    -extra_vars dbdt,topg,topg_initial,thk
done

# run into ice-free equilibrium
pisms -Lz 5000 -y 80000 -o out_equi9c.nc \
-bed_def lc -i out_equi9a.nc -Rel -150 \
-extra_file ex_equi9c.nc -extra_times 1000 \
-extra_vars dbdt,topg,topg_initial,thk

# run into icefree equilibrium with 9 restarts
cp out_equi9a.nc out_equi0d.nc
for i in {0..8}
do
    pisms -bed_def lc -i out_equi$((i))d.nc -o out_equi$((i+1))d.nc \
    -extra_file ex_equi$((i+1))d.nc -extra_times 1000 -Rel -150 -y 10000 \
    -extra_vars dbdt,topg,topg_initial,thk
done

ncrcat ex_equi[1-9]b.nc ex_equi9b2.nc
ncrcat ex_equi[1-9]d.nc ex_equi9d2.nc
```

See the [plot](http://www.pik-potsdam.de/~albrecht/pism_circular/timeseries_restart.pdf) for the test runs:

Closer look into the numbers: Final values in the center point vary only slightly for the two cases of with or withour intermediate restarts: equilibrium thk by 5m and topg by about 1m. And for the ice-free state final topg is without restart -0.08m and with 9 intermediate restarts -2.8m. Regarding the uplift by more than 1000m this appears like a rather small error...

**Comment by talbrecht** [comment]
Well -2.8 m is only true in the center of the setup, there is some funny final pattern in topg with up to 20m variance across the computational domain in the restart case (see [nc-file](http://www.pik-potsdam.de/~albrecht/pism_circular/out_equi9d.nc))... Could this still be some kind of round-off error?

**Comment by ckhroulev** [comment]
I'm sure that the code initializing the Lingle-Clark bed deformation model has a bug. I'm looking into this and will add an update once I have more information. (@talbrecht: I haven't looked at your runs, but I do have a simple setup that I can use to test this.)

**Comment by ckhroulev** [comment]
One more comment for now:

> Variable `topg_initial` is reset to `topg` at every restart (in stable v0.7.3).

This is intentional. `topg_initial` is **not** the equilibrium topography, it is what the `long_name` attribute says: the "bedrock surface elevation (at the beginning of the run)". This field is used only is the user sets `-regrid_bed_special` to adjust bed topography when switching from a coarse grid to a finer grid. In other words, `topg_initial` is not to blame here.

**Comment by ckhroulev** [comment]
Updated the issue title to reflect the real problem.

**Comment by jsegu** [comment]
I agree with [Torsten](#issuecomment-278662909). On long time scales errors become small.

Yet another example... Set `yrs` to 10000 and for relatively small errors. Set `yrs` to 1000 for significant differences between single restart (`ex{0,f}.nc`) and multiple restart (`ex{0..8}.nc`) output.

    # set variables
    args="-bed_def lc -extra_times 100 -extra_vars dbdt,thk,topg,topg_initial"
    grow="pisms $args -Rel 450"  # run pisms with high accumulation
    melt="pisms $args -Rel -150"  # run pisms with high ablation
    yrs=1000  # years between restarts, change to 10000 and problem almost gone

    # grow an ice sheet for 1000 years
    $grow -y $yrs -Lz 5000 -o out0.nc -extra_file ex0.nc

    # melt it quickly and let the bedrock rebound
    $melt -y $((8*yrs)) -i out0.nc -o outf.nc -extra_file exf.nc

    # same thing with restarts
    for i in {0..7}
    do
        $melt -y $yrs -i out$((i)).nc -o out$((i+1)).nc -extra_file ex$((i+1)).nc
    done

    # visualize all results
    ncview ex?.nc

Perhaps the error caused at initialization is only significant when the model is out of equilibrium?

**Comment by talbrecht** [comment]
This issue is also very important to me right now, I'm running deglaciation simulations for a collaborative paper and I would like to put the label "urgent" on this issue?! @ckhroulev let us know how we can help...

**Comment by jsegu** [comment]
I just re-ran my [previous example](#issuecomment-279961886) on e9d2d1f and see no more differences in topographic rebound between the restarted and one-go results. Do you think this issue is resolved? Thank you so much @ckhroulev, it looks like you nearly completely rewrote the bed deformation model!

**Comment by ckhroulev** [comment]
@jsegu This issue is resolved in the `dev` branch, but I will merge [`ckhroulev/issue-370`](https://github.com/pism/pism/tree/ckhroulev/issue-370) into `stable0.7` before I call this resolved overall.

> ... it looks like you nearly completely rewrote the bed deformation model!

The fix itself is relatively small-- most changes are related to new verification and regression tests I added to make sure this issue stays fixed.

**Comment by ckhroulev** [comment]
Closing this because it will be fixed in the next release (coming soon).

**Comment by tkleiner** [comment]
In an experiment to check if this restarting issue is fixed also for a more complex set-up than the grow/melt experiment from  https://github.com/pism/pism/issues/370#issuecomment-279961886, I run our InitMIP ctrl experiment with https://github.com/pism/pism/commit/9cfa7cb3055b1c0fbae172e9f08b457fe9af69fd for 3000 years (instead of 100 years) in one chunk and compared the output of the same set-up, but run in a sequence of 0-100,100-200,200-500,500-1000,1000-3000 years with restarts. In year 3000 topg differs by about +/- 5m.

Differences between "one-go" and "restarted" runs can be reproduced with the `circular_withshelf.nc`set-up  from `$PISM_SRC_DIR/examples/marine/circular/circular_ice_sheet.py`. This set-up runs for 3000 years within approx. 2hours on a single core (serial) on AWI's cluster. See the [run script](https://github.com/pism/pism/files/1813138/circular_bed_def_restarted.sh.txt) and the [slurm batch file](https://github.com/pism/pism/files/1813140/run.batch.txt). 

Result: pism stable1.0 and stable0.7 "one-go" runs differ by about 0.5 meter (topg) from the corresponding "restarted" runs after 3000 years.  

Details:
```
$ ncdump -h beddef_lc_1728702_outf.nc | grep "run_stats:source"
run_stats:source = "PISM stable v1.0-34-g3dfc9fa committed by Constantine Khrulev on 2018-02-28 20:46:34 -0900" ;
$ nccmp.py -v topg beddef_lc_1728702_out2.nc beddef_lc_1728702_outf.nc
Variable topg: delta = 4.296968e-01, tol = 0.000000e+00

$ ncdump -h beddef_lc_1724815_outf.nc | grep "run_stats:source"
		run_stats:source = "PISM stable v0.7.3-20-g9cfa7cb" ;
$ nccmp.py -v topg beddef_lc_1724815_out2.nc  beddef_lc_1724815_outf.nc 
    Variable topg: delta = 4.824187e-01, tol = 0.000000e+00
```  
Note, slurm ids 1728702 and 1724815 can be ignored. 

I am not sure, if I am doing something wrong and would be happy to receive your feedback.

**Comment by talbrecht** [comment]
Here's some feedback: I re-ran @tkleiner 's circular-ice-shelf experiments, see[ here](https://nbviewer.jupyter.org/url/www.pik-potsdam.de/~albrecht/notebooks/check_restartability_circular.ipynb) for plots of ice volume and max-velocity timeseries. 

I found that the "one-go" experiment and the "20-times-restarted" experiment yield very different final results, although each experiment result can be reproduced exactly (as solved in #402 ). This indicated that the restart capability in PISM (input and/or output) implies some inaccuracy that can accumulate over several restart steps. We need to check if all model-state variables necessary for an exact "restartability" are written to the restart-file, and with sufficient precision, and if they are found and read correctly...

@ckhroulev : Should we open an new issue?

**Comment by ckhroulev** [comment]
@talbrecht Regarding your last commit: yes, if it describes a different issue we should open a new issue. A comment on a closed issue is very likely to get lost (just like this one...).

**Comment by tkleiner** [comment]
@talbrecht I think the difference between the "one-go" experiment and the "20-times-restarted" experiment could also be caused by #422.

**Comment by talbrecht** [comment]
@ckhroulev: yes, makes sense, but I hope that we can close most of those issues soon ;o)
@tkleiner: no, I have rerun the same experiment with your fix from #422, but it shows the same divergence.

**Comment by ckhroulev** [comment]
@talbrecht I would love to fix as many issues as we can. Please open new issues whenever you discover new problems, though: having each issue as a separate item on the [list of issues](https://github.com/pism/pism/issues) makes it easier to keep track of them (this list provides an overview of what needs to be done).
